db/compaction/compaction_job.cc

```diff
@@ -1519,88 +1024,44 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
       RecordCompactionIOStats();
     }
 
-    // Open output file if necessary
-    if (sub_compact->builder == nullptr) {
-      status = OpenCompactionOutputFile(sub_compact);
-      if (!status.ok()) {
-        break;
-      }
-    }
-    status = sub_compact->AddToBuilder(key, value);
-    if (!status.ok()) {
-      break;
-    }
-
-    status = sub_compact->ProcessOutFlowIfNeeded(key, value);
-    if (!status.ok()) {
-      break;
-    }
-
-    const ParsedInternalKey& ikey = c_iter->ikey();
-    status = sub_compact->current_output()->meta.UpdateBoundaries(
-        key, value, ikey.sequence, ikey.type);
+    // Add current compaction_iterator key to target compaction output, if the
+    // output file needs to be close or open, it will call the `open_file_func`
+    // and `close_file_func`.
+    // TODO: it would be better to have the compaction file open/close moved
+    // into `CompactionOutputs` which has the output file information.
+    status = sub_compact->AddToOutput(
+        *c_iter.get(),
+        [this, sub_compact]() {
+          return this->OpenCompactionOutputFile(sub_compact,
+                                                sub_compact->Current());
+        },
+        [this, sub_compact, input](const Slice* next_table_min_key) {
```

Comment:
I understand that the parameter `next_table_min_key` is a raw pointer, the same as existing FinishCompactionOutputFile(). But I feel that new functions should follow the new guideline and use reference.